luigi/contrib/snowflake.py

```diff
@@ -0,0 +1,523 @@
+
+        return connection
+
+    def exists(self, connection=None):
+        """
+        Check if the task has already been run for a given update_id from the
+        marker_table
+
+        Do not re-run the task if it has been previously run
+        """
+        new_connection = connection is None
+        if new_connection:
+            connection = self.connect()
+            connection.autocommit = True
+        cursor = connection.cursor()
+        try:
+            cursor.execute("""SELECT 1 FROM {marker_table} WHERE update_id = '{update_id}' LIMIT 1""".format(marker_table=self.marker_table,
+                                                                                                             update_id=self.update_id))
+            row = cursor.fetchone()
+        except snowflake.connector.errors.ProgrammingError as e:
+            # Snowflake doesn't seem to have constants for their error code
+            if e.errno == SNOWFLAKE_GENERIC_PROGRAMMING_ERROR:
+                row = None
+            else:
+                logger.error('Snowflake error {0} ({1}): {2} ({3})'.format(e.errno, e.sqlstate, e.msg, e.sfqid))
+                raise
+
+        if new_connection:
+            connection.close()
+
+        return row is not None
+
+    def create_marker_table(self):
+        """
+        Create marker table if it doesn't exist.
+        Using a separate connection since the transaction might have to be reset.
+        """
+        connection = self.connect()
+        connection.autocommit = True
+        cursor = connection.cursor()
+        if self.use_db_timestamps:
+            sql = """CREATE TABLE {marker_table} (
+                     update_id TEXT PRIMARY KEY,
+                     target_table TEXT,
+                     inserted TIMESTAMP DEFAULT NOW())""".format(marker_table=self.marker_table)
+        else:
+            sql = """CREATE TABLE {marker_table} (
+                     update_id TEXT PRIMARY KEY,
+                     target_table TEXT,
+                     inserted TIMESTAMP);""".format(marker_table=self.marker_table)
+
+        try:
+            cursor.execute(sql)
+        except snowflake.connector.errors.ProgrammingError as e:
+            if e.errno == SNOWFLAKE_TABLE_EXISTS_ERROR:
+                pass
+            else:
+                logger.error('Snowflake error {0} ({1}): {2} ({3})'.format(e.errno, e.sqlstate, e.msg, e.sfqid))
+                raise
+
+        connection.close()
+
+
+class S3CopyToTable(rdbms.CopyToTable, _SettingsMixins):
+    """
+    Template task for inserting a data set into Snowflake from s3.
+
+    Subclass and override the required attributes:
+        * `host`,
+        * `database`,
+        * `user`,
+        * `password`,
+        * `table`,
+        * `columns`,
+        * `s3_load_path`.
+    """
+    configuration_section = 'snowflake'
+    snowflake_copy_auth = luigi.EnumParameter(enum=SnowflakeSupportedAuth, default=SnowflakeSupportedAuth.default)
```

Comment:
am i correct that the `default` (value 0) will raise an exception bc it is not currently implemented in the `SnowflakeSupportedAuth`?

Reply:
Is this a value that can be pulled from the `cfg`? I suppose someone who is using snowflake s3copytotable in multiple instances is going to be referencing the same cluster (and thus auth method) each time.